### PR TITLE
Kronolith remote organizer

### PR DESCRIPTION
**Pull request by ralflang**
Addresses https://bugs.horde.org/ticket/3965 [#3965] Track the Organizer of events

Enhances Kronolith to handle external organizers. This is mostly useful to prevent email mess with external event invitations in a horde+thunderbird scenario.

Please review

### CODE REVIEW

File: kronolith/lib/Event.php [hunk 1]
```diff
@@ -64,6 +64,13 @@
     protected $_creator = null;
 
     /**
+     * The CAL-ADDRESS of the organizer of the event, if known.
```
Comment (yunosh) [inline kronolith/lib/Event.php]: The email address...

File: kronolith/lib/Event.php [hunk 2]
```diff
@@ -775,7 +782,9 @@ public function toiCalendar($calendar)
         $vEvent->setAttribute('SUMMARY', $this->getTitle());
 
         // Organizer
-        if (count($this->attendees)) {
+        if ($this->organizer) {
+            $vEvent->setAttribute('ORGANIZER', 'mailto:' . $this->organizer, array());
+        } elseif (count($this->attendees)) {
```
Comment (yunosh) [inline kronolith/lib/Event.php]: Why elseif? The attendees should be added if they exist, not if the organizer doesn't exist.

File: kronolith/lib/Event.php [hunk 3]
```diff
@@ -1078,6 +1087,15 @@ public function fromiCalendar($vEvent, $parseAttendees = false)
             }
         } catch (Horde_Icalendar_Exception $e) {}
 
+        // Organizer
+        try {
+            $organizer = $vEvent->getAttribute('ORGANIZER');
+            if (!empty($organizer)) {
+                $this->organizer = str_replace(array('MAILTO:', 'mailto:'), '',
+                                            $organizer);
```
Comment (yunosh) [inline kronolith/lib/Event.php]: Indenting

File: kronolith/lib/Event.php [hunk 4]
```diff
@@ -1721,7 +1744,9 @@ public function toASAppointment(array $options = array())
         $message->setTimezone($this->start);
 
         // Organizer
-        if (count($this->attendees)) {
+        if ($this->organizer) {
+            $message->setOrganizer(array('email' => $this->organizer));
+        } elseif (count($this->attendees)) {
```
Comment (yunosh) [inline kronolith/lib/Event.php]: Same logic flaw like in toIcalendar().

File: kronolith/lib/Kronolith.php [hunk 5]
```diff
@@ -1909,7 +1909,14 @@ public static function sendITipNotifications(
         $view->event = $event;
         $view->imageId = $image->getContentId();
 
-        foreach ($event->attendees as $email => $status) {
+        /* Enhancement #3965: Only send updates to organizer if the user is not the organizer */
```
Comment (yunosh) [inline kronolith/lib/Kronolith.php]: No need to reference the ticket.

File: kronolith/lib/Kronolith.php [hunk 6]
```diff
@@ -1909,7 +1909,14 @@ public static function sendITipNotifications(
...
+        $receivers = $event->attendees;
+        if ($event->organizer && !$this->isUserEmail($event->creator, $event->organizer) ) {
```
Comment (yunosh) [inline kronolith/lib/Kronolith.php]: self::isUserEmail()

File: kronolith/lib/Kronolith.php [hunk 7]
```diff
@@ -1909,7 +1909,14 @@ public static function sendITipNotifications(
...
+            /* TODO: Guess organizer name and status if he is one of the attendees */
+            $receivers = array($event->organizer => array('name' => $event->organizer, 'response' => self::RESPONSE_ACCEPTED));
```
Comment (yunosh) [inline kronolith/lib/Kronolith.php]: That would send an acceptence response as soon as an attendee is editing an event that he is not the organizer of. This is not what we want.

File: kronolith/migration/23_kronolith_remote_organizer.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ * @package  Kronolith
+ */
+class KronolithRemoteOrganizer extends Horde_Db_Migration_Base
+{
+    /**
+     * Upgrade.
+     */
+    public function up()
+    {
+        $this->addColumn('kronolith_events', 'event_organizer', 'string', array('limit' => 250));
```
Comment (yunosh) [inline kronolith/migration/23_kronolith_remote_organizer.php]: Should be 255.

File: kronolith/lib/Event.php (hunk 2 above)
Comment (mrubinsk) [inline kronolith/lib/Event.php]: This conditional doesn't add attendees, it adds the current Horde user as an ORGANIZER if it is a "meeting", as determined by the fact that there are other attendees.

Comment (ralflang) [inline kronolith/lib/Event.php]: Only one Attendee is organizer. Not sure if he needs to attend at all. This means use ORGANIZER if available, fall back to CREATOR if not, have no ORGANIZER if it has no attendees (is an event but no appointment).

Comment (mrubinsk) [inline kronolith/lib/Event.php]: Agreed, I was pointing out the same :)

Comment (ralflang) [inline kronolith/lib/Event.php]: Sorry ;)

File: kronolith/lib/Kronolith.php (hunk 7 above)
Comment (ralflang) [inline kronolith/lib/Kronolith.php]: Right, reading the latest discussion on the ticket this is a problem. If ORGANIZER && CREATOR != ORGANIZER, we need to suppress all itips but attendee status changes. 
At least it does not spam other attendees. I need to read more into the spec and the code how to do it.

Comment (ralflang) [inline kronolith/lib/Kronolith.php]: Right. Changing Attendance is not an "update" but a "response". It needs its own constant and case in sendItipNotification. 

Kronolith has no UI for changing your attendance status and sending an Itip to the organizer. (While IMP allows multiple answers to the same Itip mail). Kronolith only allows to update the EVENT status. https://www.ietf.org/rfc/rfc2446.txt 2.1.1 indicates that the event status may only be edited by the organizer. 

I could ** replace** or **duplicate** the "Status" dropdown of the edit event redbox. There is still room in the row of kronolithEventStatus, kronolithEventLocation. Don't know which option is best.

Comment (mrubinsk) [inline kronolith/lib/Kronolith.php]: Quoting Ralf Lang <notifications@github.com>:

>> @@ -1909,7 +1909,14 @@ public static function sendITipNotifications(
>>          $view->event = $event;
>>          $view->imageId = $image->getContentId();
>>
>> -        foreach ($event->attendees as $email => $status) {
>> +        /* Enhancement #3965: Only send updates to organizer if  
>> the user is not the organizer */
>> +        $receivers = $event->attendees;
>> +        if ($event->organizer &&  
>> !$this->isUserEmail($event->creator, $event->organizer) ) {
>> +            /* TODO: Guess organizer name and status if he is one  
>> of the attendees */
>> +            $receivers = array($event->organizer => array('name'  
>> => $event->organizer, 'response' => self::RESPONSE_ACCEPTED));
>
> Right. Changing Attendance is not an "update" but a "response". It  
> needs its own constant and case in sendItipNotification.
>
> Kronolith has no UI for changing your attendance status and sending  
> an Itip to the organizer. (While IMP allows multiple answers to the  
> same Itip mail). Kronolith only allows to update the EVENT status.  
> https://www.ietf.org/rfc/rfc2446.txt 2.1.1 indicates that the event  
> status may only be edited by the organizer.

Probably one reason why some calendar apps will disallow changing the  
event if you are not the ORGANIZER.


> I could ** replace** or **duplicate** the "Status" dropdown of the  
> edit event redbox. There is still room in the row of  
> kronolithEventStatus, kronolithEventLocation. Don't know which  
> option is best.

I would say make the status drop down read-only and add a new field  
for response when the current user is not the ORGANIZER.

-- 
mike
The Horde Project
http://www.horde.org
https://www.facebook.com/hordeproject
https://www.twitter.com/hordeproject

Comment (yunosh) [inline kronolith/lib/Kronolith.php]: The point I was actually trying to make was that you unconditionally set the response to "accepted", while an organizer not necessarily has to attend the event.

Comment (ralflang) [inline kronolith/lib/Kronolith.php]: Yes, and reviewing this, it's even unnecessary. I use "reply" Itip Responses instead now.

**Comment by yunosh** [comment]
Can you please rebase. This merge doesn't resolve anymore.

**Comment by ralflang** [comment]
Rebasing didn't do - I manually merged the conflicts (Removing the table config from the driver and some package.xml/CHANGES foo).

**Comment by yunosh** [comment]
Parse error: syntax error, unexpected '<<' (T_SL) in /home/jan/horde-develop/kronolith/lib/Driver/Sql.php on line 502

**Comment by ralflang** [comment]
Closing this PR - I will open a new one against the rebased branch with cleaned history. This will be missing the packages.xml and CHANGES entries. I will readd them once you are OK with it.

**Pull request closed without merging**